lld/MachO/SyntheticSections.cpp

```diff
@@ -863,20 +881,42 @@ void ObjCStubsSection::setUp() {
 }
 
 uint64_t ObjCStubsSection::getSize() const {
-  return target->objcStubsFastSize * symbols.size();
+  auto stubSize = config->objcStubsMode == ObjCStubsMode::fast
+                      ? target->objcStubsFastSize
+                      : target->objcStubsSmallSize;
+  return stubSize * symbols.size();
 }
 
 void ObjCStubsSection::writeTo(uint8_t *buf) const {
   assert(in.objcSelrefs->live);
   assert(in.objcSelrefs->isFinal);
 
   uint64_t stubOffset = 0;
+  uint64_t objcMsgSendAddr;
+  uint64_t objcStubSize;
+  uint64_t objcMsgSendIndex;
+  if (config->objcStubsMode == ObjCStubsMode::fast) {
+    objcStubSize = target->objcStubsFastSize;
+    objcMsgSendAddr = in.got->addr;
+    objcMsgSendIndex = objcMsgSend->gotIndex;
+  } else {
+    assert(config->objcStubsMode == ObjCStubsMode::small);
+    objcStubSize = target->objcStubsSmallSize;
+    if (auto *d = dyn_cast<Defined>(objcMsgSend)) {
+      objcMsgSendAddr = d->getVA();
+      objcMsgSendIndex = 0;
+    } else {
+      objcMsgSendAddr = in.stubs->addr;
+      objcMsgSendIndex = objcMsgSend->stubsIndex;
+    }
+  }
```

Comment:
instead of passing `objcMsgSendAddr` and `objcMsgSendIndex`, can we just pass `objcMsgSendAddr = objbcMsgSend->getVA()`? I think that should do the right thing

more generally the current code looks like a leaky abstraction, we are wrapping `writeObjCMsgSendFastStub` and `writeObjCMsgSendSmallStub` under a common `writeObjCMsgSendStub` interface, but then we have to call `writeObjCMsgSendStub` differently anyway depending on which mode is in use. I think the above change + moving the stubs size lookup into `writeObjCMsgSendStub` will make things cleaner